elastic/kibana
# [Actionable Observability] Verify missing groups for Metric Threshold rule before scheduling no-data actions

## Summary

This PR closed #144067 by adding a function to the Metric Threshold rule to "double check" if a missing group is actually missing data or not before scheduling a "no-data" alert. This PR also changes the format of the `missingGroup` state array from a `string[]` to `Array<{ key: string, bucketKey: Record<string, string>; }>`. There is a function called `convertStringsToMissingGroupsRecord` that converts any string group by key to the new data structure by splitting on `","` for multiple group by values; the risk in this is that it will be too aggressive for group by keys with commas. If this happens, the work around is to reset that rule by changing either the KQL filter or changing the group by fields, saving and waiting for the rule to execute (once), then reverting it back; this will reset the rule state.

### Checklist

Delete any items that are not applicable to this PR.


### Testing Setup

These instructions assume you have Kibana and Elasticsearch running in a development environment

1. Clone https://github.com/elastic/high-cardinality-cluster
2. `cd high-cardinality-cluster/high_cardinality_indexer`
3. Run `DATASET="fake_hosts" EVENTS_PER_CYCLE=1 INDEX_INTERVAL=10000 ELASTICSEARCH_HOSTS=http://localhost:9200 DELAY_IN_MINUTES=3 node src/run.js`
4. Configure a Metric Threshold Rule with `document count IS BELOW 0` and group by `host.name`.
5. You will notice that every 5 minutes (minute % 5) the queue will pause for 3 minutes then start indexing again. During that 3 minute pause you should get a "no-data" alert.
6. The `no-data` alert should properly recover (even after the data back fills). Using the command above, you should expect to see a new `no-data` alert every 5 minutes and they should recover after about `126s`.

![image](https://user-images.githubusercontent.com/41702/199081134-6e3abe76-0e48-4fa8-8382-cb4568181290.png)


Personally, I would use the code on "main" to verify the alert never recovers (using the High Cardinality Indexer), then check the PR code to make sure it works by stopping Kibana, and restarting Kibana with the PR code.
